### PR TITLE
Issue 333 check sap control scope

### LINKED ISSUE #333: Check SAP Control Scope
**Extended Description**
- As a OSCAL developer , in order to check that all controls from SSP are addressed , I want to check that control-selection element inclusion is correct and matches with existing SSP controls.

See Guide_to_OSCAL-based_FedRAMP_Security_Assessment_Plans_(SAP).pdf Section 4.1

**Acceptance Criteria**
- [x] All Schematron assertion messages are declarative statements which affirm the positive test outcome.
- [x] All Schematron assertion diagnostic messages are declarative statements which explain the negative test outcome.
- [x] The Schematron code has no assertion failures when validated using `src/validations/styleguides/sch.sch` using the `advanced` phase.
- [x] XSpec unit tests for positive and negative Schematron assertion outcomes accompany all Schematron assertions (where feasible).
- [x] include-all element exists then no include-control sibling can exist.  
- [x] No include-control @control-id exist that match the values of exclude-control @control-id 
- [x] All exclude-control and include-control @control-id values exist within the associated SSP

**Story Tasks**
- [x] Tasks...

**Definition of Done**
 - [x] Acceptance criteria met
 - [x] Unit test coverage of our code > 95%
 - [x] Automated code quality checks passed
 - [x] Security reviewed and reported
 - [x] Reviewed against plain language guidelines
 - [x] Code must be self-documenting
 - [x] No local tech debt
 - [x] Load/performance tests passed – needs to be created/automated
 - [x] Documentation updated
 - [x] Architectural Decision Record completed as necessary for significant design choices
 - [ ] PR reviewed & approved
 - [ ] Source code merged

### PR DESCRIPTION
**Pull request by markXLIX**
* Checks for matches of control-inclusion *-include elements with SSP controls
* Checks that there are no duplicate [@control-id](https://github.com/control-id) values within each control-selection

Closes #333

### CODE REVIEW

**Comment by markXLIX** [comment]
Due to the problem with dist/content - this one shouldn't be merged until that is ironed out.

File: src/validations/rules/sap.sch [hunk 1]
```diff
@@ -171,6 +193,70 @@
...
+                id="duplicate-exclude-control-and-include-control-values"
+                role="error"
+                test="count($matching-control-ids) = 0">The exclude-control and include-control sibling element @control-id values must be
+                different.</sch:assert>
+            <sch:assert
+                diagnostics="duplicate-include-control-values-diagnostic"
+                id="duplicate-include-control-values"
+                role="error"
+                test="count($include-control-ids) eq count(distinct-values($include-control-ids))">The include-control/@control-id values must be
+                different.</sch:assert>
```
Comment (GaryGapinski) [inline src/validations/rules/sap.sch:230]: Suggest "The include-control/@control-id values must not be duplicated." or equivalent.

File: src/validations/rules/sap.sch [hunk 2]
```diff
@@ -171,6 +193,70 @@
...
+            <sch:assert
+                diagnostics="duplicate-exclude-control-values-diagnostic"
+                id="duplicate-exclude-control-values"
+                role="error"
+                test="count($exclude-control-ids) eq count(distinct-values($exclude-control-ids))">The exclude-control/@control-id values must be
+                different.</sch:assert>
```
Comment (GaryGapinski) [inline src/validations/rules/sap.sch:236]: Suggest "The exclude-control/@control-id values must not be duplicated." or equivalent.

File: src/validations/test/rules/sap.xspec [hunk 3]
```diff
@@ -1,6 +1,7 @@
 <?xml version="1.0" encoding="UTF-8"?>
 
 <x:description
+    run-as="external"
```
Comment (GaryGapinski) [inline src/validations/test/rules/sap.xspec:4]: Will we need this in all XSpec documents?

**Review by GaryGapinski — CHANGES_REQUESTED** [review]
The sch.sch asserts the following. Which test coverage metric is correct?
```
Excluding assertions designated as not having xspec tests, there are 23 assertions and there are 49 unit tests which reference those assertions for test coverage of 106.52%. This is informational message (not an error).
There are 26 total assertions and there are 49 unit tests which reference those assertions for test coverage of 94.23%. This is informational message (not an error).
```

File: src/validations/rules/sap.sch [hunk 4]
```diff
@@ -171,6 +193,70 @@
...
+            name="ssp-controls"
+            value="$ssp-doc//oscal:implemented-requirement/@control-id" />
+        <sch:rule
+            context="oscal:control-selection">
+            <sch:let
+                name="exclude-control-ids"
+                value="oscal:exclude-control/@control-id ! xs:string(.)" />
+            <sch:let
+                name="include-control-ids"
+                value="oscal:include-control/@control-id ! xs:string(.)" />
```
[the remark this answers is not in the document]
Comment (markXLIX) [reply src/validations/rules/sap.sch:208]: inadvertent omission.  Fixed.

File: src/validations/rules/sap.sch [hunk 5]
```diff
@@ -171,6 +193,70 @@
...
+            <sch:let
+                name="matching-control-ids"
+                value="$exclude-control-ids[. = $include-control-ids]" />
+            <sch:assert
+                diagnostics="include-all-or-include-control-diagnostic"
+                doc:guide-reference="Guide to OSCAL-based FedRAMP Security Assessment Plans (SAP) §4.1"
+                id="include-all-or-include-control"
+                role="fatal"
```
[the remark this answers is not in the document]
Comment (markXLIX) [reply src/validations/rules/sap.sch:216]: This is an XSD validation error.  So, assuming that validation was not completed prior to schematron validation, I decided that it should be considered a showstopper.

File: src/validations/rules/sap.sch [hunk 6]
```diff
@@ -171,6 +193,70 @@
...
+                test="(oscal:include-all and not(oscal:include-control)) or (oscal:include-control and not(oscal:include-all))">An OSCAL SAP control
```
[the remark this answers is not in the document]
Comment (markXLIX) [reply src/validations/rules/sap.sch:217]: I agree.  I was going on the current (at the time of making these changes) view that we would also do XSD validations when they were noted in the Guide.

File: src/validations/rules/sap.sch [hunk 7]
```diff
@@ -171,6 +193,70 @@
 
     </sch:pattern>
 
+    <sch:pattern
+        id="control-selection">
+        <sch:let
+            name="ssp-controls"
```
[the remark this answers is not in the document]
Comment (markXLIX) [reply src/validations/rules/sap.sch:199]: yes, makes sense.

File: src/validations/rules/sap.sch (hunk 1 above)
Comment (markXLIX) [reply src/validations/rules/sap.sch:230]: Sounds good.  Changed.

File: src/validations/rules/sap.sch (hunk 2 above)
Comment (markXLIX) [reply src/validations/rules/sap.sch:236]: Agreed

**Pull request merged**